src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -48,6 +51,12 @@ class PeerManagerActor(
     */
   override val maxBlacklistedNodes: Int = 32 * 8 * discoveryConfig.kademliaBucketSize
 
+  def lruCache[A, B](maxEntries: Int): mutable.Map[A, B] = new java.util.LinkedHashMap[A, B]() {
```

Comment:
probably it would be more space efficient to have something as :
```
final class LruSet[K <: AnyRef](maxSize: Long) {
  import LruSet._
  private val cache: LinkedHashMap[K, DummyObject.type] = lruCache[K, DummyObject.type](maxSize)

  def put(key: K): Unit = {
    cache.put(key, DummyObject)
  }

  def contains(key: K): Boolean = {
    cache.contains(k)
  }

  def size: Long = cache.size()
}
object LruSet {
  private case object DummyObject
}
```
That way we would not keep values unecessarily in memory, as we only call `put` and `contains`

Reply:
could you elaborate a bit more on that please? I don't quite grasp how keeping the cash in one place over the other helps with the space efficiency. Thanks

Reply:
currently we when adding nodes to cache we add always tuple `id -> node` which means that both `id` and `Node` case class are kept in memory. 

When using approach above, we would always add different `id` for each peer, but as value we would add only `DummyObject` which is singleton, which means all values in map are actually references to one object in memory, which ultimatly means that instead of multiple (with current default around 4k) values in map, we only have one.

Reply:
I found a nicer way to discard the value - using a set and keeping the map underhood
```def lruSet[A](maxEntries: Int): mutable.Set[A] =
    newSetFromMap[A](new java.util.LinkedHashMap[A, java.lang.Boolean]() {
      override def removeEldestEntry(eldest: java.util.Map.Entry[A, java.lang.Boolean]): Boolean = size > maxEntries
    }).asScala```
    
    Will resolve this thread if it looks good to you

Reply:
as for me it looks good 👍